autosklearn/ensemble_builder.py

```diff
@@ -21,33 +23,199 @@
+        precision: [16,32,64,128]
+            precision of floats to read the predictions
+        memory_limit: Optional[int]
+            memory limit in mb. If ``None``, no memory limit is enforced.
+        read_at_most: int
+            read at most n new prediction files in each iteration
+
+    Returns
+    -------
+        List[Tuple[int, float, float, float]]:
+            A list with the performance history of this ensemble, of the form
+            [[pandas_timestamp, train_performance, val_performance, test_performance], ...]
+    """
+
+    # Create the event as one of the criteria to decide if we move to he next iteration
+    event = dask.distributed.Event(event) if event else None
+
+    # The second criteria is elapsed time
+    elapsed_time = time.time() - start_time
+
+    # The last criteria is the number of iterations
+    iteration = 0
+
+    history = []
+
+    while True:
+
+        # First test for termination conditions
+        if time_left_for_ensembles < elapsed_time:
+            warnings.warn(
+                "Terminate ensemble building as not time is left (run for {}s)".format(
+                    elapsed_time
+                ),
+            )
+            break
+        if event is not None and event.is_set():
+            warnings.warn(
+                "Terminating ensemble building as SMAC process is done...")
+            break
+        if max_iterations is not None and max_iterations <= iteration:
+            warnings.warn(
+                "Terminate ensemble building because of max iterations: {} of {}".format(
+                    max_iterations,
+                    iteration
+                )
+            )
+            break
+
+        # Add the result of the run
+        # On the next while iteration, no references to
+        # ensemble builder object, so it should be garbage collected to
+        # save memory while waiting for resources
+        # Also, notice how ensemble nbest is returned, so we don't waste
+        # iterations testing if the deterministic predictions size can
+        # be fitted in memory
+        result, ensemble_nbest = EnsembleBuilder(
+            backend=backend,
+            dataset_name=dataset_name,
+            task_type=task,
+            metric=metric,
+            ensemble_size=ensemble_size,
+            ensemble_nbest=ensemble_nbest,
+            max_models_on_disc=max_models_on_disc,
+            seed=seed,
+            precision=precision,
+            memory_limit=ensemble_memory_limit,
+            read_at_most=read_at_most,
+            random_state=seed,
+        ).run(
+            time_left=time_left_for_ensembles-elapsed_time,
+            iteration=iteration,
+        )
+
+        # If pynisher kills a run, the result
+        # might be None -- so no new timestamp info
+        if result:
+            history.extend(result)
+
+        # Don;t take resources while waiting
+        dask.distributed.secede()
```

Comment:
That's cool, didn't know about this feature

Reply:
Yeah, pretty handy to have more control over the running jobs!